airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/streaming/BaseSizeEstimator.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.db.jdbc.streaming;
+
+import io.airbyte.commons.json.Jsons;
+
+public abstract class BaseSizeEstimator implements FetchSizeEstimator {
+
+  private final long bufferByteSize;
+  private final int minFetchSize;
+  private final int defaultFetchSize;
+  private final int maxFetchSize;
+
+  protected double meanByteSize = 0.0;
+
+  protected BaseSizeEstimator(final long bufferByteSize,
+                              final int minFetchSize,
+                              final int defaultFetchSize,
+                              final int maxFetchSize) {
+    this.bufferByteSize = bufferByteSize;
+    this.minFetchSize = minFetchSize;
+    this.defaultFetchSize = defaultFetchSize;
+    this.maxFetchSize = maxFetchSize;
+  }
+
+  /**
+   * Use serialized string size as an estimation of the byte size.
+   */
+  static long getEstimatedByteSize(final Object rowData) {
+    if (rowData == null) {
+      return 0L;
+    }
+    return Jsons.serialize(rowData).length() * 4L;
+  }
+
+  protected int getBoundedFetchSize() {
+    if (meanByteSize <= 0.0) {
+      return defaultFetchSize;
+    }
+    final long rawFetchSize = Math.round(bufferByteSize / meanByteSize);
+    if (rawFetchSize > Integer.MAX_VALUE) {
+      return maxFetchSize;
+    }
+    return Math.max(minFetchSize, Math.min(maxFetchSize, (int) rawFetchSize));
```

Comment:
Can this ever return 0?
What happens when the estimator estimates that even a single row is larger than available buffer?

Reply:
No, it will always return a value in the range of `[minFetchSize, maxFetchSize]`.

As I mentioned in the above comment, as long as the row can fit in the total heap, the connector can still handle it.